File: CoUML-app/Models/Collection.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using Newtonsoft.Json.Converters;

namespace CoUML_app.Models
{
	public interface ICollectionIterator<T>
	{
		public bool HasNext();
		public T GetNext();
		public bool HasPrevious();
		public T GetPrevious();
		
	}

	public interface ICollection<T>
	{
		public ICollectionIterator<T> Iterator();
		public void Insert(T item);
		public T Remove(string key);
		public T Remove(int key);
		public int size { get;}
		// public List<T> items{ get; }
		
	}
	public class GeneralCollection<T> : ICollection<T>
	{
		private List<T> _items;
		public List<T> items{
			get{
				return this._items;
			}
		}

		/// <summary>
		/// constructor
		/// </summary>
		/// <param name="items">an inital set of items</param>
		public GeneralCollection(T[] items)
		{
			this._items = new List<T>(items);
		}

		/// <summary>
		/// constructor - empty consructor
		/// </summary>
		public GeneralCollection()
		{
			this._items = new List<T>();
		}

		/// <summary>
		/// creates a collection itterator
		/// </summary>
		/// <returns>ICollectionIterator</returns>
		public ICollectionIterator<T> Iterator()
		{
			return new CollectionIterator<T>(this);
		}

		/// <summary>
		/// isnsert an item into a collection
		/// </summary>
		/// <param name="item"></param>
		public void Insert (T item)
		{
			this._items.Add(item);
		}

		/// <summary>
		/// removes item from collection
		/// </summary>
		/// <param name="itemAtIndex">index of item</param>
		/// <returns>item removed</returns>
		public T Remove(int itemAtIndex)
		{
			T item = default(T);
			if(ValidIndex(itemAtIndex))
			{
				item = this._items[itemAtIndex];
				this._items.RemoveAt(itemAtIndex);
			}
			return item;
		}

		/// <summary>
		/// removes item from collection
		/// </summary>
		/// <param name="itemAtIndex">index of item</param>
		/// <returns>item removed</returns>
		public T Remove(string itemAtIndex)
		{
			try
			{
				int i = Int32.Parse(itemAtIndex);
				return Remove(i);
			}
			catch (FormatException e)
			{
				Console.WriteLine(e.Message);
			}
			return default(T);
		}

		/// <summary>
		/// Size of the collecion
		/// </summary>
		/// <value>size</value>
		public int size
		{
			get
			{
				return this._items.Count;
			}
			
		}

		/// <summary>
		/// determins if the index number entered is in the collection
		/// </summary>
		/// <param name="i"> index </param>
		/// <returns>true if valid</returns>
		private bool ValidIndex(int i)
		{
			return i>=0 && i < size;
		}

		public T this[int i]
		{
			get {
				return this._items[i];
			}
		}
	}

	public class RelationalCollection<T> : ICollection<T> where T:UmlElement
	{
		private Dictionary<string, T> _items;
		// public List<UmlElement> items{
		// 	get{
		// 		return new List<UmlElement>(this._items.Values);
		// 	}
		// }

		public Dictionary<string, T> items{
			get{
				return _items;
			}
		}

		/// <summary>
		/// constructor
		/// </summary>
		/// <param name="collection"> set of diagram Elements</param>
		public RelationalCollection( T[] collection)
		{
			this._items = new Dictionary<string, T>();
			foreach (T item in collection)
			{
				this._items.Add(item.id, item);
			}
		}

		/// <summary>
		/// constructor - empty
		/// </summary>
		public RelationalCollection()
		{
			this._items = new Dictionary<string, T>();
		}

		public ICollectionIterator<T> Iterator()
		{
			return new CollectionIterator<T>(
				new GeneralCollection<T>(
					(new List<T>(this._items.Values)).ToArray()
				)
			);
		}
		
		public void Insert( T item)
		{
			this._items.Add( item.id, item);
		}

		public T Remove(string id)
		{
			T item = default(T);
			string guid = new Guid(id).ToString();
			if(this._items.TryGetValue(guid, out item))
				this._items.Remove(guid);
			return item;
		}

		public T Remove(int index)
		{
			T item = default(T);
			if(index >= 0 && index < size )
			{
				item = (new List<T>(this._items.Values)).ToArray()[index];
				this._items.Remove(((T)item).id);
			}
			return item;
		}


		public int size
		{
			get
			{
				return this._items.Count;
			}
		}

	}

	public class CollectionIterator<T>: ICollectionIterator<T>
	{
		private GeneralCollection<T> _collection;
		private int _position = 0;

		public CollectionIterator( GeneralCollection<T> collection)
		{
			this._collection = collection;
		}

		public bool HasNext()
		{
			return _position < _collection.size;
		}
		public T GetNext()
		{
			if(HasNext())
				return this._collection[_position++];
			throw new CollectionException("{size, position} = {" + this._collection.size + ", " + this._position + "}");
		}
		public bool HasPrevious()
		{
			return _position > 0;
		}
		public T GetPrevious()
		{
			if( HasPrevious())
				return this._collection[_position--];
			throw new CollectionException("{size, position} = {" + this._collection.size + ", " + this._position + "}");
		}
	}

	[Serializable]
	public class CollectionException: Exception
	{
		public CollectionException(string msg): base(msg){}
		protected CollectionException(System.Runtime.Serialization.SerializationInfo info,
        System.Runtime.Serialization.StreamingContext context) : base(info, context) { }
	}
	
}

File: CoUML-app/ClientApp/src/models/Collection.ts
import { Attribute } from "@angular/core";
import { Class, Component, Interface } from "./Component";
import { UmlElement, UmlElement as T , TYPE} from "./DiagramModel";
import { Operation } from "./Subcomponent";

//interface for collections to create iterators
export interface ICollectionIterator<T>
{
	hasNext():boolean;
	getNext():T;
	hasPrevious():boolean;
	getPrevious():T;
}

export interface ICollection<T>
{
	iterator(): ICollectionIterator<T>;
	insert(item: T): void;
	remove(id: any): T | null;
	removeAll(): void
	get(id: string|number): T | null;
	size: number;
	toUmlNotation(): string;
}

export class GeneralCollection<T> implements ICollection<T>{

	items: T[] = [];

	/**
	 * contstuctor with array of item type
	 * @param collection 
	 */
	constructor(items: T[])
	{
		this.items = items;
		this[TYPE] = `CoUML_app.Models.GeneralCollection\`1[[CoUML_app.Models.${typeof(items)}, CoUML-app]], CoUML-app`;
	}
	removeAll(): void {
		delete this.items;
		this.items = [];
	}

	/**
	 * Creates a collection iterator
	 * @returns CollectionIterator
	 */
	iterator(): ICollectionIterator<T>
	{
		return new CollectionIterator<T>(this);
	}

	/**
	 * isnsert an item into a collection
	 * @param item 
	 */
	insert(item: T): void {
		console.log("inserting....");
		console.log(item);
		 this.items.push(item);
	}

	/**
	 * removes item from collection
	 * @param id index of item
	 * @returns item removed
	 */
	remove(id: any): T | null
	{
		console.log(`!!!!!!!removing \n${id} \nfrom \n[${this.items.toString()}]`)
		let item: T = null;
		if(typeof(id) == 'number' && this.validIndex(id))
		{
			item = this.items[id];
			this.items = this.items.splice(id);
		}
		else
		{
			console.log("id is string");
			// let i = this.items.indexOf(id as unknown as T);
			// this.items = this.items.splice(i);
			this.items = this.items.filter(item => (item as unknown as string) !== id);
		}
		return item;
	}

	/**
	 * size of collection
	 */
	get size(): number
	{
		return this.items?.length || -1;
	}

	/**
	 * determines if the index is in the collection
	 * @param i index 
	 * @returns tue if valid index
	 */
	validIndex(i: number): boolean
	{
		try{
			return i >= 0 && i < this.size; 
		}catch(error){
			return false;
		}
	}


	// this breaks if it's not a more complex item!!!
	// get(id: string): null | T
	// {
	// 	for(let de of this.items as unknown[] as DiagramElement[])
	// 		if(id == de?.id)
	// 				return de as unknown as T;
	// 	return null;
	// }
	get(i: number):null | T
	{
		return this.validIndex(i)? this.items[i]: null;
	}

	toUmlNotation(): string {
		return this.items.toString();
	}
}

/**
 * RelationshipCollection
 */
export class RelationalCollection<T extends UmlElement> implements ICollection<T>{

	private items: Map<string, T> = new Map<string, T>();
	// get $type() {
	// 	let ttype = T.name == Operation.name ? "Operation": 
	// 		T.name == Attribute.name ? "Attribute": "UmlElement";
			
	// 	// (this.size>0? this.items.values().next().value[TYPE]:"CoUML_app.Models.UmlElement, CoUML-app")
	// 	// return `CoUML_app.Models.RelationalCollection\`1[["CoUML_app.Models.UmlElement, CoUML-app"]], CoUML-app`
	// };
	toJSON(): any {
		return {
			typeName: "CoUML_app.Models.RelationalCollection\`1[[CoUML_app.Models.UmlElement, CoUML-app]], CoUML-app",
			items: Object.fromEntries(this.items),
		}
	 }

	constructor(collection: T[])
	{
		for(let elem of collection)
			this.insert(elem);
		// this[TYPE] = "CoUML_app.Models.RelationalCollection, CoUML-app";
	}
	removeAll(): void {
		delete this.items;
		this.items = new Map<string, T>();
	}


	iterator(): ICollectionIterator<T> 
	{
		return new CollectionIterator<T>(
				new GeneralCollection<T>(
					Array.from(this.items.values())
				)
			);
	}

	insert(item: T): void {
		console.log("inserting....");
		console.log(item);
		this.items.set(item.id, item);
	}

	remove(id: string): T | null 
	{
		console.log(`removing....${id}`);
		let removedItem = null
		if(this.items.has(id))
		{
			removedItem = this.items.get(id);
			this.items.delete(id);
		}
		return removedItem;
	}
	
	get(id: string): null | T
	{
		if(this.items.has(id))
			return this.items.get(id);

		return this.find(id);
	}

	private find(key: string): null | T
	{
		// for(let e of this.items.values())
		// {
		// 	if(!(e instanceof RelationalCollection)){
		// 		let de = e as Component;
		// 		if(key === de.name)
		// 			return de;
		// 	}
		// }
		return null;
	}

	get size(): number 
	{
		return this.items.size
	}

	toUmlNotation(): string {

		let uml: string = "";
		this.items.forEach((item:T, id: string)=>{
			uml += (uml != "" ? ", " : "") + item.toUmlNotation();
		});
		
		return uml;
	}

}

/**
 * AttributeCollectionIterator
 * fetches objects from collection
 */
 export class CollectionIterator<T> implements ICollectionIterator<T>{
	_collection: GeneralCollection<T>;
	_position: number = 0;

	constructor( collection: GeneralCollection<T> )
	{
		this._collection = collection;
	}

	hasNext():boolean{
		return  this._position < this._collection.size;
	}
	getNext():T{
		if(this.hasNext){
			return this._collection.items[this._position++];
		}
	}
	hasPrevious():boolean{
		return this._position > 0;
	}
	getPrevious():T{
		if(this.hasPrevious){
			return this._collection.items[this._position--];
		}
	}
}
